go/host/enclave/service.go

```diff
@@ -200,24 +186,85 @@ func (e *Service) Unsubscribe(id rpc.ID) error {
+// We don't need to worry about edge cases at this point, if something is wrong we will fix it in the
+// ensureActiveSequencer loop, but on the happy path this is much smoother
+func (e *Service) loadActiveSequencer() {
+	for _, guardian := range e.enclaveGuardians {
+		if guardian.state.IsEnclaveActiveSequencer() {
+			e.logger.Info("Found active sequencer enclave", log.EnclaveIDKey, guardian.GetEnclaveID())
+			e.activeSequencerID.Store(guardian.GetEnclaveID())
+			return
+		}
+	}
+}
+
+// ensureActiveSequencer is a background goroutine that attempts to ensure there is always exactly one
+// active sequencer enclave. It will promote a new active sequencer if the current one fails or is unavailable.
+func (e *Service) ensureActiveSequencer() {
+	for e.running.Load() {
+		activeSeqID := e.activeSequencerID.Load()
+		newSequencerRequired := true // this will be set false if we find the expected active sequencer is healthy
+
+		// do a pass through all guardians to see if they are in their expected state
 		for _, guardian := range e.enclaveGuardians {
-			enclID := guardian.GetEnclaveID()
-			e.logger.Info("Attempting to promote new sequencer.", log.EnclaveIDKey, enclID)
-			err := guardian.PromoteToActiveSequencer()
-			if err != nil {
-				e.logger.Info("Unable to promote new sequencer.", log.EnclaveIDKey, enclID, log.ErrKey, err)
-				continue
+			if guardian.state.IsEnclaveActiveSequencer() {
+				if guardian.GetEnclaveID().Hex() == activeSeqID.Hex() {
+					// active seq ID is set, and there is a matching guardian whose enclave reports it is the active sequencer
+					if guardian.IsSequencerOperational() {
+						// this is the expected state, so we can skip the rest of the checks
+						e.logger.Trace("Found active sequencer operational", log.EnclaveIDKey, activeSeqID)
+						newSequencerRequired = false
+					} else {
+						// sequencer is not operational, so we need to promote a new one
+						e.logger.Warn("Found active sequencer in guardians but not operational", log.EnclaveIDKey, activeSeqID)
+						guardian.DemoteFromActiveSequencer()
+					}
+				} else { // guardian thinks it is the active sequencer, but it is not the expected one
+					// demoting it will restart it and clear the active sequencer flag on the enclave.
+					// We don't have to worry about two batch producers, the guardian always checks it is expected to
+					// be the active seq before producing.
+					e.logger.Warn("Found unexpected active sequencer in guardians, demoting",
+						log.EnclaveIDKey, guardian.GetEnclaveID(), "expectedActiveID", activeSeqID.Hex())
+					guardian.DemoteFromActiveSequencer()
+				}
 			}
-			e.activeSequencerID.Store(enclID)
-			e.logger.Warn("Successfully promoted new sequencer.", log.EnclaveIDKey, enclID)
-			return
 		}
-		// wait for retry interval before trying again, enclaves may not be ready yet or may be awaiting permissioning
-		time.Sleep(_promoteSeqRetryInterval)
+
+		if newSequencerRequired {
```

Comment:
if all this logic was in a function: `promoteNewActiveEnclave`, you could call it directly where you set the flag